### PR TITLE
[Feature] fix spirte, label test when EGL context lost

### PR DESCRIPTION
**Pull request by Mee-gu**
fix spirte, label test when EGL context lost

### CODE REVIEW

File: cocos/renderer/backend/ProgramCache.cpp [hunk 1]
```diff
@@ -121,10 +121,22 @@ void ProgramCache::removeUnusedProgram()
...
 {
-    for (auto& program : _cachedPrograms) {
+    for (auto& program : _cachedPrograms)
+    {
         program.second->release();
     }
     _cachedPrograms.clear();
 }
 
+void ProgramCache::reloadDefaultProgram()
```
Comment (minggo) [inline cocos/renderer/backend/ProgramCache.cpp]: default is not correct, it includes programs set outside

File: cocos/renderer/backend/Program.h [hunk 2]
```diff
@@ -26,6 +26,12 @@ class Program : public Ref
     virtual int getMaxFragmentLocation() const = 0;
     virtual const std::unordered_map<std::string, AttributeBindInfo> getActiveAttributes() const = 0;
 
+#if CC_ENABLE_CACHE_TEXTURE_DATA
+    virtual void reloadGLProgram() = 0;
+    virtual void releaseGLProgram() = 0;
+    virtual int getMappedLocation(int location) const = 0;
+#endif
```
Comment (minggo) [inline cocos/renderer/backend/Program.h]: these functions don't have to be public, only need `reloadProgram()`

Comment (minggo) [reply cocos/renderer/backend/Program.h]: `getMappedLocation ` is not meaningful

File: cocos/renderer/backend/ProgramState.cpp [hunk 3]
```diff
@@ -404,7 +409,14 @@ void ProgramState::setTexture(int location, uint32_t slot, backend::Texture* tex
     info.slot = {slot};
     info.textures = {texture};
     info.retainTextures();
+
+#if CC_ENABLE_CACHE_TEXTURE_DATA
+    auto mappedLocation = _program->getMappedLocation(location);
+    textureInfo.erase(location);
+    textureInfo[mappedLocation] = std::move(info);
```
Comment (minggo) [inline cocos/renderer/backend/ProgramState.cpp]: Should erase all textureInfo once when EGL context lost, not every frame.

File: cocos/renderer/backend/ProgramState.cpp [hunk 4]
```diff
@@ -414,7 +426,13 @@ void ProgramState::setTextureArray(int location, const std::vector<uint32_t>& sl
     info.slot = slots;
     info.textures = textures;
     info.retainTextures();
+#if CC_ENABLE_CACHE_TEXTURE_DATA
+    auto mappedLocation = _program->getMappedLocation(location);
+    textureInfo.erase(location);
+    textureInfo[mappedLocation] = std::move(info);
```
Comment (minggo) [inline cocos/renderer/backend/ProgramState.cpp]: the same issue as above

File: cocos/renderer/backend/opengl/ShaderModuleGL.h [hunk 5]
```diff
@@ -11,15 +11,19 @@ class ShaderModuleGL : public ShaderModule
...
     inline GLuint getShader() const { return _shader; }
     
 private:
     void compileShader(ShaderStage stage, const std::string& source);
     char* getErrorLog(GLuint shader) const;
-    void deleteShader();
+
+#ifdef CC_ENABLE_CACHE_TEXTURE_DATA
+    virtual void deleteShader() override ;
+#endif
```
Comment (minggo) [inline cocos/renderer/backend/opengl/ShaderModuleGL.h]: Why need to add macro here?

File: cocos/renderer/CCRenderer.cpp [hunk 6]
```diff
@@ -201,6 +201,17 @@ void Renderer::init()
...
+
+#if CC_ENABLE_CACHE_TEXTURE_DATA
+    _cacheTextureListener = EventListenerCustom::create(EVENT_RENDERER_RECREATED, [this](EventCustom* event){
+        /** listen the event that renderer was recreated on Android/WP8 */
+        _triangleCommandBufferManager.init();
+        _vertexBuffer = _triangleCommandBufferManager.getVertexBuffer();
+        _indexBuffer = _triangleCommandBufferManager.getIndexBuffer();
+    });
+
+    Director::getInstance()->getEventDispatcher()->addEventListenerWithFixedPriority(_cacheTextureListener, -1);
```
Comment (minggo) [inline cocos/renderer/CCRenderer.cpp]: Should handle buffer reload in backend.

File: cocos/renderer/CCTexture2D.cpp [hunk 7]
```diff
@@ -341,6 +347,23 @@ bool Texture2D::initWithMipmaps(MipmapInfo* mipmaps, int mipmapsNum, PixelFormat
...
+        VolatileTextureMgr::setSamplerDescriptor(this, texParams);
+    }
+    else
+    {
+        TexParams texParams = {_hasMipmaps,
+                               backend::SamplerFilter ::NEAREST, backend::SamplerFilter ::NEAREST, backend::SamplerFilter ::NEAREST,
+                               backend::SamplerAddressMode ::DONT_CARE, backend::SamplerAddressMode ::DONT_CARE};
+        VolatileTextureMgr::setSamplerDescriptor(this, texParams);
+    }
+#endif
```
Comment (minggo) [inline cocos/renderer/CCTexture2D.cpp:347]: Can we handle these params inside backend texture itself?

**Comment by Mee-gu** [comment]
update review

File: cocos/renderer/CCTexture2D.cpp [hunk 8]
```diff
@@ -374,8 +377,12 @@ bool Texture2D::initWithMipmaps(MipmapInfo* mipmaps, int mipmapsNum, PixelFormat
     CCASSERT(textureDescriptor.textureFormat != backend::TextureFormat::NONE, "TextureFormat should not be NONE");
     
     textureDescriptor.compressed = info.compressed;
-    
-    _texture = static_cast<backend::Texture2D*>(device->newTexture(textureDescriptor));
+
+    if(_texture == nullptr)
+    {
+        CC_SAFE_RELEASE(_texture);
```
Comment (minggo) [inline cocos/renderer/CCTexture2D.cpp]: release a null texture?

File: cocos/renderer/backend/opengl/BufferGL.cpp [hunk 9]
```diff
@@ -34,6 +59,11 @@ void BufferGL::updateData(void* data, unsigned int size)
         CHECK_GL_ERROR_DEBUG();
         _bufferAllocated = size;
     }
+
+#if CC_ENABLE_CACHE_TEXTURE_DATA
+    if(BufferUsage::STATIC ==  _usage)
+        BufferManager::addBufferData(this, data, size);
+#endif
```
Comment (minggo) [inline cocos/renderer/backend/opengl/BufferGL.cpp:80]: why not just record the data itself?

**Comment by Mee-gu** [comment]
update review

File: cocos/renderer/backend/opengl/ProgramGL.cpp [hunk 10]
```diff
@@ -17,6 +20,21 @@ ProgramGL::ProgramGL(const std::string& vertexShader, const std::string& fragmen
...
+#if CC_ENABLE_CACHE_TEXTURE_DATA
+    for(const auto& uniform: _uniformInfos)
+    {
+        UniformLocation uniformLocation;
+        uniformLocation.location = glGetUniformLocation(_program, uniform.first.c_str());
+        _originalUniformLocations[uniform.first] = uniformLocation;
+        _uniformLocationMap[uniform.second.location] = uniform.second.location;
+    }
+
+    _backToForegroundListener = EventListenerCustom::create(EVENT_RENDERER_RECREATED, [this](EventCustom*){
```
Comment (minggo) [inline cocos/renderer/backend/opengl/ProgramGL.cpp:33]: where to remove the listener?

File: cocos/renderer/backend/opengl/ProgramGL.cpp [hunk 11]
```diff
@@ -185,6 +217,13 @@ void ProgramGL::computeUniformInfos()
...
     UniformLocation uniformLocation;
+#if CC_ENABLE_CACHE_TEXTURE_DATA
+    if(_originalUniformLocations.find(uniform) != _originalUniformLocations.end())
+        return _originalUniformLocations.at(uniform);
+    else
+        return uniformLocation;
+#endif
+
     uniformLocation.location = glGetUniformLocation(_program, uniform.c_str());
     return uniformLocation;
```
Comment (minggo) [inline cocos/renderer/backend/opengl/ProgramGL.cpp:233]: it is better to put these these codes into `#else section`

File: cocos/renderer/backend/opengl/ProgramGL.cpp (hunk 10 above)
Comment (Mee-gu) [reply cocos/renderer/backend/opengl/ProgramGL.cpp:33]: I forget to remove the listener, and it will be removed in the destructor.

**Comment by Mee-gu** [comment]
update review

**Review by minggo — APPROVED** [review]

**Pull request merged**